### PR TITLE
liquidSVM

### PR DESCRIPTION
**Pull request by PhilippPro**
This is the implementation of liquidSVM as mlr-learner. The probability predictions are still technically flawed and they are not (yet) responding to my emails, so unfortunately probability predictions are not provided at the moment.

### CODE REVIEW

**Comment by larskotthoff** [comment]
Could you add some tests please?

**Comment by PhilippPro** [comment]
Why are there some errors on the pr and not on the push? I also do not understand the errors very well...

**Comment by larskotthoff** [comment]
Ok, it looks like the tests are failing with a segmentation fault. I've tried on two different machines and can't reproduce it. Does it work when you run the tests locally?

**Comment by PhilippPro** [comment]
Yes, they were ok. The algorithm only produces warnings in the default version which is not good. Also the probability prediction is not working properly yet. But the package has quite potential regarding performance and runtime and I already contacted the creators.

File: R/RLearner_classif_liquidSVM.R [hunk 1]
```diff
@@ -0,0 +1,49 @@
+#' @export
+makeRLearner.classif.liquidSVM = function() {
+  makeRLearnerClassif(
+    cl = "classif.liquidSVM",
+    package = "liquidSVM",
+    par.set = makeParamSet(
+      makeIntegerLearnerParam(id = "d", lower = 0L),
+      makeLogicalLearnerParam(id = "scale"),
```
Comment (larskotthoff) [inline R/RLearner_classif_liquidSVM.R]: Default for this is TRUE.

File: R/RLearner_classif_liquidSVM.R [hunk 2]
```diff
@@ -0,0 +1,49 @@
+#' @export
+makeRLearner.classif.liquidSVM = function() {
+  makeRLearnerClassif(
+    cl = "classif.liquidSVM",
+    package = "liquidSVM",
+    par.set = makeParamSet(
+      makeIntegerLearnerParam(id = "d", lower = 0L),
```
Comment (larskotthoff) [inline R/RLearner_classif_liquidSVM.R]: This should not be tunable (`tunable = FALSE`).

File: R/RLearner_classif_liquidSVM.R [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+      makeLogicalLearnerParam(id = "scale"),
+      makeIntegerLearnerParam(id = "threads", lower = -1L),
```
Comment (larskotthoff) [inline R/RLearner_classif_liquidSVM.R]: Default is 0.

File: R/RLearner_classif_liquidSVM.R [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+      makeIntegerLearnerParam(id = "partition_choice", lower = 0L),
```
Comment (larskotthoff) [inline R/RLearner_classif_liquidSVM.R]: Default is 0.

File: R/RLearner_classif_liquidSVM.R [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+      makeIntegerLearnerParam(id = "grid_choice", lower = -2L, upper = 2L),
```
Comment (larskotthoff) [inline R/RLearner_classif_liquidSVM.R]: How does this interact with `gamma_steps`, `lambda_steps`, etc? Are these options mutually exclusive?

File: R/RLearner_classif_liquidSVM.R [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+      makeIntegerLearnerParam(id = "adaptivity_control", lower = 0L, upper = 2L),
```
Comment (larskotthoff) [inline R/RLearner_classif_liquidSVM.R]: Default is 0.

File: R/RLearner_classif_liquidSVM.R [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+      makeIntegerLearnerParam(id = "random_seed", default = 1),
```
Comment (larskotthoff) [inline R/RLearner_classif_liquidSVM.R]: I didn't find any documentation on what the default random seed is -- where did you find this? It also shouldn't be tunable.

File: R/RLearner_classif_liquidSVM.R (hunk 5 above)
Comment (larskotthoff) [reply R/RLearner_classif_liquidSVM.R]: And all of these options should be not tunable.

File: R/RLearner_classif_liquidSVM.R [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+      makeIntegerLearnerParam(id = "fold"),
```
Comment (larskotthoff) [inline R/RLearner_classif_liquidSVM.R]: Should that be `folds`? And there should be a minimum for this (and again not tunable).

File: R/RLearner_classif_liquidSVM.R [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+      makeIntegerLearnerParam(id = "clipping"),
```
Comment (larskotthoff) [inline R/RLearner_classif_liquidSVM.R]: Seems that the minimum is -1. Also probably shouldn't be tunable.

File: R/RLearner_classif_liquidSVM.R [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+      makeIntegerLearnerParam(id = "gamma_steps", lower = 0),
+      makeNumericLearnerParam(id = "min_gamma", lower = 0),
+      makeNumericLearnerParam(id = "max_gamma", lower = 0),
+      makeNumericVectorLearnerParam(id = "gammas", lower = 0),
+      makeIntegerLearnerParam(id = "lambda_steps", lower = 0),
+      makeNumericLearnerParam(id = "min_lambda", lower = 0),
+      makeNumericLearnerParam(id = "max_lambda", lower = 0),
+      makeNumericVectorLearnerParam(id = "lambdas", lower = 0),
+      makeNumericVectorLearnerParam(id = "c_values", lower = 0),
+      makeLogicalLearnerParam(id = "useCells")
```
Comment (larskotthoff) [inline R/RLearner_classif_liquidSVM.R]: Seems that the default for this is `FALSE`. Should be not tunable.

File: R/RLearner_regr_liquidSVM.R [hunk 11]
```diff
@@ -0,0 +1,43 @@
+#' @export
+makeRLearner.regr.liquidSVM = function() {
+  makeRLearnerRegr(
+    cl = "regr.liquidSVM",
+    package = "liquidSVM",
+    par.set = makeParamSet(
```
Comment (larskotthoff) [inline R/RLearner_regr_liquidSVM.R]: See comments for `classif` param set.

**Comment by PhilippPro** [comment]
I tried to change all parameters according to your suggestions. I am unsure about clipping, if it is tunable or not, that's why I did not set `tunable = FALSE`.

**Comment by larskotthoff** [comment]
Thanks, looks good -- what about restrictions between parameters that control the tuning though? If you set all the gamma parameters at the same time, I guess that some of them override others? It may make sense to add requirements saying that if you set a certain parameter, others cannot be set to avoid surprising results when users accidentally set conflicting parameters.

**Comment by PhilippPro** [comment]
> Thanks, looks good -- what about restrictions between parameters that control the tuning though? If you set all the gamma parameters at the same time, I guess that some of them override others? It may make sense to add requirements saying that if you set a certain parameter, others cannot be set to avoid surprising results when users accidentally set conflicting parameters.

I changed the parameter settings to take into account requirements. Moreover I looked into https://github.com/cran/liquidSVM/blob/master/R/mlr.R and changed the parameter settings slightly.

**Comment by larskotthoff** [comment]
Thanks for all your work on this. Merging.

**Pull request merged**